// @loaders.gl/images/dist/es5/image-loader.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _parseImage = _interopRequireDefault(require("./lib/parsers/parse-image"));

var _binaryImageParsers = require("./lib/binary-image-api/binary-image-parsers");

var VERSION = typeof "2.0.0" !== 'undefined' ? "2.0.0" : 'latest';
var EXTENSIONS = ['png', 'jpg', 'jpeg', 'gif', 'webp', 'bmp', 'ico', 'svg'];
var MIME_TYPES = ['image/png', 'image/jpeg', 'image/gif', 'image/webp', 'image/bmp', 'image/vndmicrosofticon', 'image/svg+xml'];
var ImageLoader = {
  name: 'Images',
  version: VERSION,
  mimeTypes: MIME_TYPES,
  extensions: EXTENSIONS,
  parse: _parseImage["default"],
  test: function test(arrayBuffer) {
    var dataView = new DataView(arrayBuffer);
    return (0, _binaryImageParsers.isJpeg)(dataView) || (0, _binaryImageParsers.isBmp)(dataView) || (0, _binaryImageParsers.isGif)(dataView) || (0, _binaryImageParsers.isPng)(dataView);
  },
  options: {
    image: {
      format: 'auto',
      decode: true
    }
  }
};
var _default = ImageLoader;
exports["default"] = _default;
