// @loaders.gl/images/dist/es5/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getSupportedImageType = getSupportedImageType;
Object.defineProperty(exports, "ImageLoader", {
  enumerable: true,
  get: function get() {
    return _imageLoader["default"];
  }
});
Object.defineProperty(exports, "HTMLImageLoader", {
  enumerable: true,
  get: function get() {
    return _imageLoader["default"];
  }
});
Object.defineProperty(exports, "ImageWriter", {
  enumerable: true,
  get: function get() {
    return _imageWriter["default"];
  }
});
Object.defineProperty(exports, "isImageTypeSupported", {
  enumerable: true,
  get: function get() {
    return _imageType.isImageTypeSupported;
  }
});
Object.defineProperty(exports, "isImage", {
  enumerable: true,
  get: function get() {
    return _parsedImageApi.isImage;
  }
});
Object.defineProperty(exports, "getImageType", {
  enumerable: true,
  get: function get() {
    return _parsedImageApi.getImageType;
  }
});
Object.defineProperty(exports, "getImageSize", {
  enumerable: true,
  get: function get() {
    return _parsedImageApi.getImageSize;
  }
});
Object.defineProperty(exports, "getImageData", {
  enumerable: true,
  get: function get() {
    return _parsedImageApi.getImageData;
  }
});
Object.defineProperty(exports, "isBinaryImage", {
  enumerable: true,
  get: function get() {
    return _binaryImageApi.isBinaryImage;
  }
});
Object.defineProperty(exports, "getBinaryImageMIMEType", {
  enumerable: true,
  get: function get() {
    return _binaryImageApi.getBinaryImageMIMEType;
  }
});
Object.defineProperty(exports, "getBinaryImageSize", {
  enumerable: true,
  get: function get() {
    return _binaryImageApi.getBinaryImageSize;
  }
});
Object.defineProperty(exports, "loadImage", {
  enumerable: true,
  get: function get() {
    return _loadImage.loadImage;
  }
});
Object.defineProperty(exports, "loadImageArray", {
  enumerable: true,
  get: function get() {
    return _loadImageArray.loadImageArray;
  }
});
Object.defineProperty(exports, "loadImageCube", {
  enumerable: true,
  get: function get() {
    return _loadImageCube.loadImageCube;
  }
});
Object.defineProperty(exports, "_JPEGLoader", {
  enumerable: true,
  get: function get() {
    return _imageLoaders.JPEGLoader;
  }
});
Object.defineProperty(exports, "_PNGLoader", {
  enumerable: true,
  get: function get() {
    return _imageLoaders.PNGLoader;
  }
});
Object.defineProperty(exports, "_GIFLoader", {
  enumerable: true,
  get: function get() {
    return _imageLoaders.GIFLoader;
  }
});
Object.defineProperty(exports, "_BMPLoader", {
  enumerable: true,
  get: function get() {
    return _imageLoaders.BMPLoader;
  }
});
Object.defineProperty(exports, "_SVGLoader", {
  enumerable: true,
  get: function get() {
    return _imageLoaders.SVGLoader;
  }
});
Object.defineProperty(exports, "_ImageLoaders", {
  enumerable: true,
  get: function get() {
    return _imageLoaders.ImageLoaders;
  }
});

var _imageLoader = _interopRequireDefault(require("./image-loader"));

var _imageWriter = _interopRequireDefault(require("./image-writer"));

var _imageType = require("./lib/parsed-image-api/image-type");

var _parsedImageApi = require("./lib/parsed-image-api/parsed-image-api");

var _binaryImageApi = require("./lib/binary-image-api/binary-image-api");

var _loadImage = require("./lib/texture-loading-api/load-image");

var _loadImageArray = require("./lib/texture-loading-api/load-image-array");

var _loadImageCube = require("./lib/texture-loading-api/load-image-cube");

var _imageLoaders = require("./image-loaders");

function getSupportedImageType(imageType) {
  return (0, _imageType.getDefaultImageType)(imageType);
}
