// @loaders.gl/images/dist/es5/lib/parsed-image-api/parsed-image-api.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isImage = isImage;
exports.deleteImage = deleteImage;
exports.getImageType = getImageType;
exports.getImageSize = getImageSize;
exports.getImageData = getImageData;

var _typeof2 = _interopRequireDefault(require("@babel/runtime/helpers/typeof"));

var _assert = _interopRequireDefault(require("../utils/assert"));

function isImage(image) {
  return Boolean(getImageTypeOrNull(image));
}

function deleteImage(image) {
  switch (getImageType(image)) {
    case 'imagebitmap':
      image.close();
      break;

    default:
  }
}

function getImageType(image) {
  var format = getImageTypeOrNull(image);

  if (!format) {
    throw new Error('Not an image');
  }

  return format;
}

function getImageSize(image) {
  switch (getImageType(image)) {
    case 'imagebitmap':
    case 'ndarray':
      return {
        width: image.width,
        height: image.height
      };

    case 'html':
      return {
        width: image.naturalWidth,
        height: image.naturalHeight
      };

    default:
      return (0, _assert["default"])(false);
  }
}

function getImageData(image) {
  switch (getImageType(image)) {
    case 'imagebitmap':
    case 'html':
      var canvas = document.createElement('canvas');
      var context = canvas.getContext('2d');
      canvas.width = image.width;
      canvas.height = image.height;
      context.drawImage(image, 0, 0);
      var imageData = context.getImageData(0, 0, image.width, image.height);
      return imageData.data;

    case 'ndarray':
      return image.data;

    default:
      return (0, _assert["default"])(false);
  }
}

function getImageTypeOrNull(image) {
  if (typeof ImageBitmap !== 'undefined' && image instanceof ImageBitmap) {
    return 'imagebitmap';
  }

  if (typeof Image !== 'undefined' && image instanceof Image) {
    return 'html';
  }

  if (image && (0, _typeof2["default"])(image) === 'object' && image.data && image.width && image.height) {
    return 'ndarray';
  }

  if (Array.isArray(image)) {
    return 'ndarray';
  }

  return null;
}
