// @deck.gl/extensions/dist/es6/path-style/path-style.js
import { LayerExtension, _mergeShaders as mergeShaders } from '@deck.gl/core';
import { dashShaders } from './shaders.glsl';
const defaultProps = {
  getDashArray: {
    type: 'accessor',
    value: [0, 0]
  },
  dashJustified: false
};
export default class PathStyleExtension extends LayerExtension {
  constructor({
    dash = false
  } = {}) {
    super({
      dash
    });
  }

  isEnabled(layer) {
    return layer.state.pathTesselator;
  }

  getShaders(extension) {
    if (!extension.isEnabled(this)) {
      return null;
    }

    let result = {};

    if (extension.opts.dash) {
      result = mergeShaders(result, dashShaders);
    }

    return result;
  }

  initializeState(context, extension) {
    const attributeManager = this.getAttributeManager();

    if (!attributeManager || !extension.isEnabled(this)) {
      return;
    }

    extension.enabled = true;

    if (extension.opts.dash) {
      attributeManager.add({
        instanceDashArrays: {
          size: 2,
          accessor: 'getDashArray'
        }
      });
    }
  }

  updateState(params, extension) {
    if (!extension.isEnabled(this)) {
      return;
    }

    const uniforms = {};

    if (extension.opts.dash) {
      uniforms.dashAlignMode = this.props.dashJustified ? 1 : 0;
    }

    this.state.model.setUniforms(uniforms);
  }

}
PathStyleExtension.extensionName = 'PathStyleExtension';
PathStyleExtension.defaultProps = defaultProps;
