// @loaders.gl/images/dist/es6/lib/parsed-image-api/parsed-image-api.js
import assert from '../utils/assert';
export function isImage(image) {
  return Boolean(getImageTypeOrNull(image));
}
export function deleteImage(image) {
  switch (getImageType(image)) {
    case 'imagebitmap':
      image.close();
      break;

    default:
  }
}
export function getImageType(image) {
  const format = getImageTypeOrNull(image);

  if (!format) {
    throw new Error('Not an image');
  }

  return format;
}
export function getImageSize(image) {
  switch (getImageType(image)) {
    case 'imagebitmap':
    case 'ndarray':
      return {
        width: image.width,
        height: image.height
      };

    case 'html':
      return {
        width: image.naturalWidth,
        height: image.naturalHeight
      };

    default:
      return assert(false);
  }
}
export function getImageData(image) {
  switch (getImageType(image)) {
    case 'imagebitmap':
    case 'html':
      const canvas = document.createElement('canvas');
      const context = canvas.getContext('2d');
      canvas.width = image.width;
      canvas.height = image.height;
      context.drawImage(image, 0, 0);
      const imageData = context.getImageData(0, 0, image.width, image.height);
      return imageData.data;

    case 'ndarray':
      return image.data;

    default:
      return assert(false);
  }
}

function getImageTypeOrNull(image) {
  if (typeof ImageBitmap !== 'undefined' && image instanceof ImageBitmap) {
    return 'imagebitmap';
  }

  if (typeof Image !== 'undefined' && image instanceof Image) {
    return 'html';
  }

  if (image && typeof image === 'object' && image.data && image.width && image.height) {
    return 'ndarray';
  }

  if (Array.isArray(image)) {
    return 'ndarray';
  }

  return null;
}
