// @loaders.gl/images/dist/es6/index.js
export { default as ImageLoader } from './image-loader';
export { default as ImageWriter } from './image-writer';
export { isImageTypeSupported } from './lib/parsed-image-api/image-type';
export { isImage, getImageType, getImageSize, getImageData } from './lib/parsed-image-api/parsed-image-api';
export { isBinaryImage, getBinaryImageMIMEType, getBinaryImageSize } from './lib/binary-image-api/binary-image-api';
export { loadImage } from './lib/texture-loading-api/load-image';
export { loadImageArray } from './lib/texture-loading-api/load-image-array';
export { loadImageCube } from './lib/texture-loading-api/load-image-cube';
export { JPEGLoader as _JPEGLoader, PNGLoader as _PNGLoader, GIFLoader as _GIFLoader, BMPLoader as _BMPLoader, SVGLoader as _SVGLoader, ImageLoaders as _ImageLoaders } from './image-loaders';
export { default as HTMLImageLoader } from './image-loader';
import { getDefaultImageType } from './lib/parsed-image-api/image-type';
export function getSupportedImageType(imageType) {
  return getDefaultImageType(imageType);
}
